File: Leetcode/Graphs/133. Clone Graph.cpp
class Solution {
private:
    unordered_map <Node*, Node*> Map;
public:
    Node* cloneGraph(Node* node) {
        if (node) {
            if (Map.find(node) != Map.end()) return Map[node];
            Map[node] = new Node(node -> val);
            for (Node* Neighbor : node -> neighbors) 
                Map[node] -> neighbors.push_back(cloneGraph(Neighbor));
            return Map[node];
        }
        return node;
    }
};

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

File: Leetcode/Graphs/133. Clone Graph.java
class Solution {
    private HashMap <Node, Node> Map = new HashMap<>();
    public Node cloneGraph(Node node) {
        if (node != null) {
            if (Map.containsKey(node)) return Map.get(node);
            Node temp = new Node(node.val);
            Map.put(node, temp);
            for (Node Neighbor : node.neighbors)
                temp.neighbors.add(cloneGraph(Neighbor));
            return temp;
        }
        return node;
    }
}
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/